eva_submission/biosamples_submission.py

```diff
@@ -0,0 +1,471 @@
+#!/usr/bin/env python
+# Copyright 2020 EMBL - European Bioinformatics Institute
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+from csv import DictReader, DictWriter
+from datetime import datetime
+
+import requests
+from cached_property import cached_property
+from ebi_eva_common_pyutils.config import cfg
+from ebi_eva_common_pyutils.logger import AppLogger
+
+from eva_submission.xlsx.xlsx_parser_eva import EVAXLSReader
+
+_now = datetime.now().isoformat()
+
+
+class HALNotReadyError(Exception):
+    pass
+
+
+class HALCommunicator(AppLogger):
+    """
+    This class helps navigate through REST API that uses the HAL standard.
+    """
+    acceptable_code = [200, 201]
+
+    def __init__(self, aap_url, bsd_url, username, password):
+        self.aap_url = aap_url
+        self.bsd_url = bsd_url
+        self.username = username
+        self.password = password
+
+    def _validate_response(self, response):
+        """Check that the response has an acceptable code and raise if it does not"""
+        if response.status_code not in self.acceptable_code:
+            self.error(response.request.method + ': ' + response.request.url + " with " + str(response.request.body))
+            self.error("headers: {}".format(response.request.headers))
+            self.error("<{}>: {}".format(response.status_code, response.text))
+            raise ValueError('The HTTP status code ({}) is not one of the acceptable codes ({})'.format(
+                str(response.status_code), str(self.acceptable_code))
+            )
+        return response
+
+    @cached_property
+    def token(self):
+        """Retrieve the token from the AAP REST API then cache it for further quering"""
+        response = requests.get(self.aap_url, auth=(self.username, self.password))
+        self._validate_response(response)
+        return response.text
+
+    def _req(self, method, url, **kwargs):
```

Comment:
I compared this to this [other file](https://github.com/EBIvariation/metadata/blob/3bc04293bc2508060fc9b54ad27b9e92d3158bba/submission_to_biosamples/submit_to_bsd.py#L69) and the retry decorator does not seem to be there for the method.

Reply:
That's correct. I removed it when testing but it should be there.